File: go-shorturl/internal/api/ratelimit.lua
-- File: middleware/ratelimit.lua
-- KEYS[1] = The unique key (e.g., rate_limit:127.0.0.1)
-- ARGV[1] = Burst Capacity (e.g., 10 tokens)
-- ARGV[2] = Refill Rate (e.g., 1 token per second)
-- ARGV[3] = Current Unix Time (Seconds)
-- ARGV[4] = Tokens requested (usually 1)

local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local rate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local requested = tonumber(ARGV[4])

-- Get current state (tokens and last_refill_time)
local bucket = redis.call("hmget", key, "tokens", "last_refill")
local tokens = tonumber(bucket[1])
local last_refill = tonumber(bucket[2])

-- Initialize if key doesn't exist
if not tokens then
    tokens = capacity
    last_refill = now
end

-- Calculate refill
-- (now - last_refill) * rate = new tokens to add
local delta = math.max(0, now - last_refill)
local filled_tokens = math.min(capacity, tokens + (delta * rate))

-- Check if allowed
local allowed = 0
if filled_tokens >= requested then
    filled_tokens = filled_tokens - requested
    allowed = 1
    
    -- Save new state
    redis.call("hmset", key, "tokens", filled_tokens, "last_refill", now)
    -- Expire key after 60s of inactivity to save RAM
    redis.call("expire", key, 60)
end

return allowed

File: go-shorturl/internal/api/ratelimit.go
package api

import (
	_ "embed" // Standard library to embed files
	"fmt"
	"net/http"
	"time"

	"github.com/redis/go-redis/v9"
)

// Embed the lua script into the binary so we don't need the file at runtime
//
//go:embed ratelimit.lua
var limitScript string

type RateLimiterConfig struct {
	Client *redis.Client
	Limit  int // Max burst
	Rate   int // Tokens/sec refilled
}

func RateLimitMiddleware(cfg RateLimiterConfig) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			ctx := r.Context()

			// 1. Identify User (IP)
			// In production, use "X-Forwarded-For" if behind a proxy
			ip := r.RemoteAddr
			key := fmt.Sprintf("rate_limit:%s", ip)

			// 2. Prepare Arguments for Lua
			now := time.Now().Unix()

			// 3. Fire the Atomic Script
			// Returns 1 (Allowed) or 0 (Denied)
			result, err := cfg.Client.Eval(ctx, limitScript, []string{key},
				cfg.Limit, // ARGV[1]
				cfg.Rate,  // ARGV[2]
				now,       // ARGV[3]
				1,         // ARGV[4] (tokens requested)
			).Result()

			if err != nil {
				// Fail Open: If Redis is down, let traffic through
				// Log the error in real life
				next.ServeHTTP(w, r)
				return
			}

			// 4. Handle Result
			if result.(int64) == 0 {
				w.WriteHeader(http.StatusTooManyRequests)
				w.Write([]byte("429 - Too Many Requests (Token Bucket Exhausted)\n"))
				return
			}

			next.ServeHTTP(w, r)
		})
	}
}
